### PR TITLE
Backend improvements

### PR DESCRIPTION
**Pull request by logan-anderson**
A couple of small(ish) Backend improvements
- Previously when using with a backend other then next.js one would have to fake. req.query.routes. Now that is handled. 
-  Add ability to pass a base path other then /api/tina/ (Currently will not work with NextAuth Default Session provider )

### CODE REVIEW

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 1]
```diff
@@ -35,22 +30,61 @@ export const LocalBackendAuthProvider = () =>
...
   const { initialize, isAuthorized, extraRoutes } = authProvider
   initialize?.().catch((e) => {
     console.error(e)
   })
+  // setup options
+  let basePath = options?.basePath || '/api/tina/'
+  // ensure basePath has trailing slash
+  basePath = basePath.endsWith('/') ? basePath : `${basePath}/`
+  // ensure basePath has leading slash
+  basePath = basePath.startsWith('/') ? basePath : `/${basePath}`
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:78]:
```suggestion
  const basePath = options?.basePath ? `/${options.basePath.replace(/^\/?/, '').replace(/\/?$/, '')}/` : '/api/tina/'
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 2]
```diff
@@ -59,29 +93,44 @@ function MakeNodeApiHandler({
...
+    const url = new URL(path, `http://${req.headers?.host || 'localhost'}`)
+
+    // Remove the basePath from the url
+    const routes = url.pathname?.replace(opts.basePath, '')?.split('/')
+
     if (typeof routes === 'string') {
       throw new Error('Please name your next api route [...routes] not [route]')
     }
-    if (!routes?.length) {
+    if (!routes?.length || routes.length === 0) {
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:114]: I think this is redundant
```suggestion
    if (!routes?.length) {
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 3]
```diff
@@ -59,29 +93,44 @@ function MakeNodeApiHandler({
...
       gql: {
-        handler: async (...params) => {
-          const [req, res] = params
+        handler: async (req, res, _opts) => {
           if (req.method !== 'POST') {
             res.statusCode = 405
-            res.write(JSON.stringify({ error: 'method not allowed' }))
+            res.write(
+              JSON.stringify({
+                error: 'Method not allowed. Can only post to /gql',
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:131]:
```suggestion
                error: 'Method not allowed. Only POST requests are supported by /gql',
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 4]
```diff
@@ -35,22 +30,61 @@ export const LocalBackendAuthProvider = () =>
...
+     * @default /api/tina
+     */
+    basePath?: string
+  }
 }
+export type NodeApiHandler = (
+  req: IncomingMessage,
+  res: ServerResponse
+) => Promise<void>
+
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:57]:
```suggestion

type NodeRouterHandlerOptions = Required<TinaBackendOptions['options']>

```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 5]
```diff
@@ -35,22 +30,61 @@ export const LocalBackendAuthProvider = () =>
...
+type NodeRouteHandler = (
+  req: IncomingMessage,
+  res: ServerResponse,
+  opts: Required<TinaBackendOptions['options']>
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:61]:
```suggestion
  opts: NodeRouterHandlerOptions
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 6]
```diff
@@ -35,22 +30,61 @@ export const LocalBackendAuthProvider = () =>
...
     console.error(e)
   })
+  // setup options
+  let basePath = options?.basePath || '/api/tina/'
+  // ensure basePath has trailing slash
+  basePath = basePath.endsWith('/') ? basePath : `${basePath}/`
+  // ensure basePath has leading slash
+  basePath = basePath.startsWith('/') ? basePath : `/${basePath}`
+
+  const opts: Required<TinaBackendOptions['options']> = {
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:80]:
```suggestion
  const opts: NodeRouterHandlerOptions = {
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 7]
```diff
@@ -59,29 +93,44 @@ function MakeNodeApiHandler({
   isAuthorized,
   extraRoutes,
   databaseClient,
-}: BackendAuthProvider & { databaseClient: DatabaseClient }) {
-  const handler: NodeApiHandler = async (...params) => {
-    const [req, res] = params
+  opts,
+}: BackendAuthProvider & {
+  databaseClient: DatabaseClient
+  opts: Required<TinaBackendOptions['options']>
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:99]:
```suggestion
  opts: NodeRouterHandlerOptions
```

File: packages/tinacms-authjs/src/index.ts [hunk 8]
```diff
@@ -144,12 +144,24 @@ const AuthJsBackendAuthProvider = ({
...
+          // The domain is not important here, we just need to parse the pathName
+          const url = new URL(
+            req.url,
+            `http://${req.headers?.host || 'localhost'}`
+          )
+
+          // basePath always has leading and trailing slash
+          const prefix = `${opts.basePath}auth/`
+          // get everything in the path after `${basePath}auth/`
+          const everythingAfterAuth = url.pathname
```
Comment (kldavis4) [inline packages/tinacms-authjs/src/index.ts:158]:
```suggestion
          const authSubRoutes = url.pathname
```

File: packages/tinacms-authjs/src/index.ts [hunk 9]
```diff
@@ -144,12 +144,24 @@ const AuthJsBackendAuthProvider = ({
...
+            ?.replace(prefix, '')
+            ?.split('/')
+
+          // This is required for NextAuth to work properly
           // @ts-ignore
-          req.query.nextauth = rest
+          req.query.nextauth = everythingAfterAuth
```
Comment (kldavis4) [inline packages/tinacms-authjs/src/index.ts:164]:
```suggestion
          req.query.nextauth = authSubRoutes
```

File: packages/tinacms-authjs/src/index.ts [hunk 10]
```diff
@@ -144,12 +144,24 @@ const AuthJsBackendAuthProvider = ({
...
+          const url = new URL(
+            req.url,
+            `http://${req.headers?.host || 'localhost'}`
+          )
+
+          // basePath always has leading and trailing slash
+          const prefix = `${opts.basePath}auth/`
+          // get everything in the path after `${basePath}auth/`
+          const everythingAfterAuth = url.pathname
+            ?.replace(prefix, '')
```
Comment (kldavis4) [inline packages/tinacms-authjs/src/index.ts:159]:
```suggestion
            ?.replace(`${opts.basePath}auth/`, '') // basePath always has leading and trailing slash
```

File: packages/tinacms-authjs/src/index.ts [hunk 11]
```diff
@@ -144,12 +144,24 @@ const AuthJsBackendAuthProvider = ({
...
+
+        handler: async (req, res, opts) => {
+          // The domain is not important here, we just need to parse the pathName
+          const url = new URL(
+            req.url,
+            `http://${req.headers?.host || 'localhost'}`
+          )
+
+          // basePath always has leading and trailing slash
+          const prefix = `${opts.basePath}auth/`
```
Comment (kldavis4) [inline packages/tinacms-authjs/src/index.ts:156]:
```suggestion
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 12]
```diff
@@ -35,22 +30,58 @@ export const LocalBackendAuthProvider = () =>
...
+     */
+    basePath?: string
+  }
 }
+export type NodeApiHandler = (
+  req: IncomingMessage,
+  res: ServerResponse
+) => Promise<void>
+
+type NodeRouterHandlerOptions = Required<TinaBackendOptions['options']>
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:58]:
```suggestion
type NodeRouteHandlerOptions = Required<TinaBackendOptions['options']>
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 13]
```diff
@@ -35,22 +30,58 @@ export const LocalBackendAuthProvider = () =>
...
+
+type NodeRouteHandler = (
+  req: IncomingMessage,
+  res: ServerResponse,
+  opts: NodeRouterHandlerOptions
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:63]:
```suggestion
  opts: NodeRouteHandlerOptions
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 14]
```diff
@@ -35,22 +30,58 @@ export const LocalBackendAuthProvider = () =>
...
   databaseClient,
+  options,
 }: TinaBackendOptions) {
   const { initialize, isAuthorized, extraRoutes } = authProvider
   initialize?.().catch((e) => {
     console.error(e)
   })
+  const basePath = options?.basePath ? `/${options.basePath.replace(/^\/?/, '').replace(/\/?$/, '')}/` : '/api/tina/'
+
+  const opts: NodeRouterHandlerOptions = {
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:77]:
```suggestion
  const opts: NodeRouteHandlerOptions = {
```

File: packages/@tinacms/datalayer/src/backend/index.ts [hunk 15]
```diff
@@ -59,29 +90,44 @@ function MakeNodeApiHandler({
   isAuthorized,
   extraRoutes,
   databaseClient,
-}: BackendAuthProvider & { databaseClient: DatabaseClient }) {
-  const handler: NodeApiHandler = async (...params) => {
-    const [req, res] = params
+  opts,
+}: BackendAuthProvider & {
+  databaseClient: DatabaseClient
+  opts: NodeRouterHandlerOptions
```
Comment (kldavis4) [inline packages/@tinacms/datalayer/src/backend/index.ts:96]:
```suggestion
  opts: NodeRouteHandlerOptions
```

**Comment by logan-anderson** [comment]
@kldavis4 ok if I merge this?

**Review by kldavis4 — APPROVED** [review]

**Pull request merged**